cpp/src/arrow/compute/exec/asof_join_node.cc

```diff
@@ -951,28 +953,27 @@ class AsofJoinNode : public ExecNode {
   }
 
   static arrow::Result<std::shared_ptr<Schema>> MakeOutputSchema(
-      const std::vector<ExecNode*>& inputs,
+      const std::vector<std::shared_ptr<Schema>> input_schema,
       const std::vector<col_index_t>& indices_of_on_key,
       const std::vector<std::vector<col_index_t>>& indices_of_by_key) {
     std::vector<std::shared_ptr<arrow::Field>> fields;
 
-    size_t n_by = indices_of_by_key[0].size();
+    size_t n_by = indices_of_by_key.size() == 0 ? 0 : indices_of_by_key[0].size();
```

Comment:
When can `indices_of_by_key.size() == 0` means this is a join without `by` key?

Reply:
Yes, this happens (and I observed it before adding the code change here) when the by-key is empty.